ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PullQueryStreamWriter.java

```diff
@@ -0,0 +1,369 @@
+    // Interrupt the writing thread in case it's blocking for another row.
+    final Thread thread = writingThread.get();
+    if (thread != null && Thread.currentThread() != thread) {
+      thread.interrupt();
+    }
+  }
+
+  /**
+   * Converts the object to json and returns the string.
+   * @param object The object to convert
+   * @return The serialized JSON
+   */
+  private String writeValueAsString(final Object object) {
+    try {
+      return objectMapper.writeValueAsString(object);
+    } catch (final JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Converts the KsqlNode to KsqlHostInfoEntity
+   */
+  private static Optional<KsqlHostInfoEntity> toKsqlHostInfo(final Optional<KsqlNode> ksqlNode) {
+    return ksqlNode.map(
+        node -> new KsqlHostInfoEntity(node.location().getHost(), node.location().getPort()));
+  }
+
+  /**
+   * State that's kept for the buffered response and the last flush time.
+   */
+  private static class WriterState {
+    private final Clock clock;
+    // The buffer of JSON that we're always flushing as we hit either time or size thresholds.
+    private StringBuilder sb = new StringBuilder();
+    // Last flush timestamp in millis
+    private long lastFlushMs;
+
+    WriterState(final Clock clock) {
+      this.clock = clock;
+    }
+
+    public WriterState append(final String str) {
+      sb.append(str);
+      return this;
+    }
+
+    public int length() {
+      return sb.length();
+    }
+
+    public long getLastFlushMs() {
+      return lastFlushMs;
+    }
+
+    public String getStringToFlush() {
+      final String str = sb.toString();
+      sb = new StringBuilder();
+      lastFlushMs = clock.millis();
+      return str;
+    }
+  }
+
+  /**
+   * Wraps the PullQueryQueue to keep a hold of the head of the queue explicitly so it always knows
+   * if there's something next.
+   */
+  private static final class QueueWrapper {
+    private final PullQueryQueue pullQueryQueue;
+    private final long disconnectCheckInterval;
+    // We always keep a reference to the head of the queue so that we know if there's another
+    // row in the result in order to produce proper JSON.
+    private PullQueryRow head = null;
+
+    QueueWrapper(final PullQueryQueue pullQueryQueue, final long disconnectCheckInterval) {
+      this.pullQueryQueue = pullQueryQueue;
+      this.disconnectCheckInterval = disconnectCheckInterval;
+    }
+
+    public boolean hasAnotherRow() {
+      return head != null;
+    }
+
+    public PullQueryRow pollNextRow() throws InterruptedException {
+      final PullQueryRow row = pullQueryQueue.pollRow(
+          disconnectCheckInterval,
+          TimeUnit.MILLISECONDS
+      );
+
+      if (row != null) {
+        // The head becomes the next thing to process and the newly polled row becomes the head.
+        final PullQueryRow toProcess = head;
+        head = row;
+        return toProcess;
```

Comment:
i guess your algorithm above handles this gracefully, but the first time you call `pollNextRow` won't it always be `null`? might just want to add a little comment there as well so that the next person doesn't scratch their head 😄

Reply:
Yes, that's right.  I'll add a comment.